site/src/pages/WorkspacePage/WorkspaceScheduleControls.tsx

```diff
@@ -48,28 +44,50 @@ export const WorkspaceScheduleControls: FC<WorkspaceScheduleControlsProps> = ({
   );
   const deadlinePlusEnabled = maxDeadlineIncrease >= 1;
   const deadlineMinusEnabled = maxDeadlineDecrease >= 1;
-
-  const onDeadlineChangeSuccess = () => {
-    displaySuccess("Updated workspace shutdown time.");
-  };
-  const onDeadlineChangeFails = (error: unknown) => {
-    displayError(
-      getErrorMessage(error, "Failed to update workspace shutdown time."),
+  const deadlineUpdateTimeout = useRef<number>();
+  const lastStableDeadline = useRef<Dayjs>(deadline);
+
+  const updateWorkspaceDeadlineQueryData = (deadline: Dayjs) => {
+    queryClient.setQueryData(
+      workspaceByOwnerAndNameKey(workspace.owner_name, workspace.name),
+      {
+        ...workspace,
+        latest_build: {
+          ...workspace.latest_build,
+          deadline: deadline.toISOString(),
+        },
+      },
     );
   };
-  const decreaseMutation = useMutation({
-    ...decreaseDeadline(workspace),
-    onSuccess: onDeadlineChangeSuccess,
-    onError: onDeadlineChangeFails,
-  });
-  const increaseMutation = useMutation({
-    ...increaseDeadline(workspace),
-    onSuccess: onDeadlineChangeSuccess,
-    onError: onDeadlineChangeFails,
+
+  const updateDeadlineMutation = useMutation({
+    ...updateDeadline(workspace),
+    onSuccess: (_, updatedDeadline) => {
+      displaySuccess("Workspace shutdown time has been successfully updated.");
+      lastStableDeadline.current = updatedDeadline;
+    },
+    onError: (error) => {
+      displayError(
+        getErrorMessage(
+          error,
+          "We couldn't update your workspace shutdown time. Please try again.",
+        ),
+      );
+      updateWorkspaceDeadlineQueryData(lastStableDeadline.current);
+    },
   });
 
+  const handleDeadlineChange = (newDeadline: Dayjs) => {
+    clearTimeout(deadlineUpdateTimeout.current);
+    // Optimistic update
+    updateWorkspaceDeadlineQueryData(newDeadline);
+    deadlineUpdateTimeout.current = window.setTimeout(() => {
+      updateDeadlineMutation.mutate(newDeadline);
+    }, 500);
+  };
+
   return (
-    <div css={styles.scheduleValue}>
+    <div css={styles.scheduleValue} data-testid="schedule-controls">
```

Comment:
Wondering if this could be turned into a `<fieldset>` for better semantics. It's usually used in a `<form>`, but I  *think* it can be used by itself, too, mainly to help group related controls

Reply:
IMO, it does not make too much sense to use a fieldset outside of a form. 

> The <fieldset> HTML element is used to group several controls as well as labels (<label>) within a web form.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset